### ISSUE TITLE
@-links 404 if an user changes their name

### ISSUE THREAD

**Issue by nonamethanks** [danbooru/danbooru#4681]
This is mainly an issue with the moderation report (https://danbooru.donmai.us/forum_topics/16565?page=34#forum_post_179255 the reported user changed their name so it's 404 now) but it's also a problem with quotes in comments, forum etc, where the old name is displayed instead of the new. 

A proper way to fix this issue is to make it so that writing @-user links to their user ID, rather than to the name search (right now it links to <https://danbooru.donmai.us/users?name=username>). 
Quotes should also link to the user account being quoted rather than just write "x said:", otherwise it becomes really hard to follow old forum threads where users with deleted accounts posted.

It's also possible that an old username link will redirect to a different user if someone changes their name and another user takes the old name.

**Comment by evazion** [comment]
This would basically require rewriting `@evazion` to `<@#52664>` when the comment is saved, then rendering it as `@evazion` when the comment is displayed. This is fine, except it would make editing slightly harder, since if you write someone's name then go back to edit your post, now it's been replaced by some random id (edit: I suppose we could do `<@evazion#52664>` so that the name is still there).

The bigger problem is doing this correctly. The easy way is to mangle the comment using regexes without actually parsing it, but that won't handle things like `[code]` or `[nodtext]` blocks correctly. The correct way is parse it to HTML, rewrite the mention, then convert it back to DText, but then the syntax may change because parsing isn't round-trippable. For example, you can write links like `"link":[http://google.com]` or `[link](http://google.com)`, but we don't keep track of which syntax you used, so when we convert it back to DText what you wrote may go from `[link](http://google.com)` to `"link":[http://google.com]`. Other examples include `[[cat]]s` vs `[[cat|cats]]`, `[i]` vs `<em>`, etc.

There are tons of cases where there are multiple ways to write the same thing, and keeping track of that would basically require completely rewriting the DText parser. The easy-but-wrong way is the only real option here.

Even with regexes, the rules are a bit tricky:

https://github.com/danbooru/danbooru/blob/af12a1e649cfc2afe04f9eafaf4266a19cd87119/lib/dtext_rb/ext/dtext/dtext.cpp.rl#L151-L166

https://github.com/danbooru/danbooru/blob/af12a1e649cfc2afe04f9eafaf4266a19cd87119/lib/dtext_rb/test/test_dtext.rb#L69-L208
